File: rust-api/report/src/models/axe.rs
use prelude::models::report_model::{CreateReport, ReportLevel, ReportType};
use serde::{Deserialize, Serialize};

#[derive(Clone, Deserialize, Serialize)]

enum ViolationImpact {
    #[serde(rename = "serious")]
    Serious,
    #[serde(rename = "moderate")]
    Moderate,
    #[serde(rename = "minor")]
    Minor,
    #[serde(rename = "critical")]
    Critical,
}

impl From<ViolationImpact> for ReportLevel {
    fn from(violation_impact: ViolationImpact) -> Self {
        match violation_impact {
            ViolationImpact::Serious => ReportLevel::Error,
            ViolationImpact::Critical | ViolationImpact::Moderate => ReportLevel::Warning,
            ViolationImpact::Minor => ReportLevel::Info,
        }
    }
}

#[derive(Deserialize, Serialize)]
struct Node {
    html: String,
}

#[derive(Deserialize, Serialize)]
struct Violation {
    impact: ViolationImpact,
    help: String,
    #[serde(rename = "helpUrl")]
    help_url: String,
    nodes: Vec<Node>,
}

#[derive(Deserialize, Serialize)]
pub struct Axe {
    violations: Vec<Violation>,
}

impl From<&Violation> for Vec<CreateReport> {
    fn from(violation: &Violation) -> Self {
        violation
            .nodes
            .iter()
            .map(move |node| CreateReport {
                report_type: ReportType::A11y,
                level: violation.impact.clone().into(),
                title: violation.help.to_owned(),
                context: node.html.to_owned(),
                help: Some(violation.help_url.to_owned()),
            })
            .collect()
    }
}

impl From<&Axe> for Vec<CreateReport> {
    fn from(axe: &Axe) -> Self {
        axe.violations
            .iter()
            .flat_map::<Vec<CreateReport>, _>(|v| v.into())
            .collect()
    }
}

File: go-api/frontendmentor/models/axe.go
package models

type node struct {
	Html string `json:"html"`
}

type violation struct {
	Impact  string `json:"impact"`
	Help    string `json:"help"`
	HelpURL string `json:"helpUrl"`
	Nodes   []node `json:"nodes,omitempty"`
}

func (v *violation) ToIssues() *[]Issue {
	var issues []Issue

	for _, node := range v.Nodes {

		Level := impactToLevel(&v.Impact)
		Title := v.Help
		Context := node.Html
		Help := &v.HelpURL

		if Level == nil || Title == "" || Context == "" || Help == nil {
			continue
		}

		issue := Issue{
			Title:   Title,
			Level:   *Level,
			Context: Context,
			Help:    Help,
		}

		issues = append(issues, issue)
	}
	return &issues
}

type Axe struct {
	Violations []violation `json:"violations,omitempty"`
}

func (a *Axe) ToIssues() *[]Issue {
	var issues []Issue

	for _, v := range a.Violations {
		if len(v.Nodes) == 0 {
			continue
		}

		issues = append(issues, *v.ToIssues()...)
	}

	return &issues
}

func impactToLevel(impact *string) *string {
	var result string
	switch *impact {
	case "serious":
		result = "error"
	case "critical", "moderate":
		result = "warning"
	default:
		result = "info"
	}

	return &result
}
